### PR TITLE
FEATURE: Integrate conflict resolution with publish/discard dialog workflow

### LINKED ISSUE #3761: FEATURE: Offer rebase option when publish/discard fails due to conflict
If a user logs in and finds their workspace in `OUTDATED` or  `OUTDATED_CONFLICT` state, they cannot publish any changes, before they have rebased their workspace.

image

It is possible however, that a user's workspace transitions into  `OUTDATED` or `OUTDATED_CONFLICT` state in the background without the UI noticing that transition. The user is then able to use all options from the Publish Dropdown. If they attempt to publish their changes, they will receive an exception:
Screenshot_2024-03-25_14-16-37 Publish Error

If such a case occurs, the UI should offer an option to rebase the workspace immediately and then re-attempt the publishing operation.

Here's what an integrated workflow could look like:
Untitled-2024-03-26-1434 Integrated Rebase Workflow
(https://excalidraw.com/#room=fe18ec088dcbdc396eed,KxgT9UbEp0yXJVtS3vjV5Q)

**Open Questions**

- [ ] Coming from the Publish Dropdown, users are acting in the scope of either the current site or the current document. The "rebase" operation however is global to the current content repository. How do we communicate that?

**Acceptance Criteria**
- [ ] When users click "Publish", "Publish All", "Discard" or "Discard All" while their workspace is in `OUTDATED` or `OUTDATED_CONFLICT` state, they are presented with a dialog that allows them to rebase their workspace immediately
- [ ] When users click on "Synchronize now" in said dialog, the rebase workflow as established in https://github.com/neos/neos-ui/issues/3744 starts
- [ ] Once the rebase workflow is finished successfully, the initial "Publish", "Publish All", "Discard" or "Discard All" operation is automatically re-attempted
- [ ] There is an E2E test for at least one of the above scenarios
- [ ] The UI refreshes `ContentCanvas` and `DocumentTree` after rebase

### LINKED ISSUE #3785: BUG: The new conflict resolution e2e tests are sadly a bit flaky in ci :O
The new e2e tests are sadly a bit flaky in ci :O

_Originally posted by @mhsdesign in https://github.com/neos/neos-ui/issues/3762#issuecomment-2111068584_
            

```
 Syncing
 ✖ Syncing: Create a conflict state between two editors and choose "Discard all" as a resolution strategy during rebase

   1) - Error while restoring configuration after Role switch -
      A JavaScript error occurred on "http://onedimension.localhost:8081/neos/content?node=user-editor__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__f676459d-ca77-44bc-aeea-44114814c279".
      Repeat test actions in the browser and check the console for errors.
      Enable the “skipJsErrors” option to ignore JavaScript errors during test execution. Learn more: "https://testcafe.io/documentation/404038/recipes/debugging/skip-javascript-errors"
      If the website only throws this error when you test it with TestCafe, please create a new issue at:
      "https://github.com/DevExpress/testcafe/issues/new?template=bug-report.md".

      JavaScript error details:
      Error: NetworkError when attempting to fetch resource.
      generalErrorHandler@http://onedimension.localhost:8081/_Resources/Static/Packages/Neos.Neos.Ui/Build/Host.js?b403a802:14:155902
      Lye</zye</m</<@http://onedimension.localhost:8081/_Resources/Static/Packages/Neos.Neos.Ui/Build/Host.js?b403a802:15:28034

      Browser: Firefox 126.0 / Windows 10 (https://app.saucelabs.com/tests/65e9848a30964224968563e3504567e3)

         138 |            timeout: 30000
         139 |        });
         140 |}
         141 |
         142 |async function switchToRole(t, role) {
       > 143 |    await t.useRole(role);
         144 |    await waitForReact(30000);
         145 |    await Page.goToPage('Home');
         146 |}
         147 |
         148 |async function chooseDiscardAllAndFinishSynchronization(t) {

         at <anonymous> (/home/circleci/app/Packages/Application/Neos.Neos.Ui/Tests/IntegrationTests/Fixtures/1Dimension/syncing.e2e.js:143:13)
         at asyncGeneratorStep (/home/circleci/app/Packages/Application/Neos.Neos.Ui/Tests/IntegrationTests/Fixtures/1Dimension/syncing.e2e.js:7:28)
         at _next (/home/circleci/app/Packages/Application/Neos.Neos.Ui/Tests/IntegrationTests/Fixtures/1Dimension/syncing.e2e.js:7:28)
         at <anonymous> (/home/circleci/app/Packages/Application/Neos.Neos.Ui/Tests/IntegrationTests/Fixtures/1Dimension/syncing.e2e.js:7:28)
         at <anonymous> (/home/circleci/app/Packages/Application/Neos.Neos.Ui/Tests/IntegrationTests/Fixtures/1Dimension/syncing.e2e.js:7:28)
         at switchToRole (/home/circleci/app/Packages/Application/Neos.Neos.Ui/Tests/IntegrationTests/Fixtures/1Dimension/syncing.e2e.js:142:28)
         at <anonymous> (/home/circleci/app/Packages/Application/Neos.Neos.Ui/Tests/IntegrationTests/Fixtures/1Dimension/syncing.e2e.js:128:11)
         at asyncGeneratorStep (/home/circleci/app/Packages/Application/Neos.Neos.Ui/Tests/IntegrationTests/Fixtures/1Dimension/syncing.e2e.js:7:28)
         at _next (/home/circleci/app/Packages/Application/Neos.Neos.Ui/Tests/IntegrationTests/Fixtures/1Dimension/syncing.e2e.js:7:28)


 ✖ Syncing: Create a conflict state between two editors and choose "Drop conflicting changes" as a resolution strategy during rebase

   1) The specified selector does not match any element in the DOM tree.

       > | Selector('#neos-workspace-rebase')

      Browser: Firefox 126.0 / Windows 10 (https://app.saucelabs.com/tests/65e9848a30964224968563e3504567e3)

         67 |    await switchToRole(t, editorUserOnOneDimensionTestSite);
         68 |
         69 |    //
         70 |    // Sync changes from "admin"
         71 |    //
       > 72 |    await t.click(Selector('#neos-workspace-rebase'));
         73 |    await t.click(Selector('#neos-SyncWorkspace-Confirm'));
         74 |    await t.wait(1000);
         75 |
         76 |    //
         77 |    // Assert that all 3 documents are now visible in the document tree

         at <anonymous> (/home/circleci/app/Packages/Application/Neos.Neos.Ui/Tests/IntegrationTests/Fixtures/1Dimension/syncing.e2e.js:72:13)
         at asyncGeneratorStep (/home/circleci/app/Packages/Application/Neos.Neos.Ui/Tests/IntegrationTests/Fixtures/1Dimension/syncing.e2e.js:7:28)
         at _next (/home/circleci/app/Packages/Application/Neos.Neos.Ui/Tests/IntegrationTests/Fixtures/1Dimension/syncing.e2e.js:7:28)
```

### LINKED ISSUE #3845: Log if publish failed?
@nezaniel said:

> it seems that exceptions like "ContentStream is closed" (and especially the previous publishing problem) do not get logged.

Because we catch them currently:

https://github.com/neos/neos-ui/blob/40d20cd32f0bf553be02ef341b5c2b66d837634b/Classes/Controller/BackendServiceController.php#L223

slack https://neos-project.slack.com/archives/C04PYL8H3/p1724660274447629

### PR DESCRIPTION
**Pull request by grebaldi**
https://github.com/neos/neos-ui/assets/2522299/5177f375-a586-4c97-8511-a9b60fd4b7a9

solves: https://github.com/neos/neos-ui/issues/3761
fixes: https://github.com/neos/neos-ui/issues/3845
fixes: #3785
builds on: https://github.com/neos/neos-ui/pull/3762

### The Problem

With the introduction of conflict resolution, users can synchronize their workspace with recent changes in the base workspace by clicking the "sync"-button that is placed aside the publish dropdown. When a workspace is in `OUTDATED` state, the "sync"-button shows up.

It may happen though that a workspace transitions into `OUTDATED` state without the UI taking notice of it. If the user then tries to publish their changes, publishing fails with an exception. Only after reloading the entire UI, they'll be able to trigger the "sync"-workflow.

**Steps to reproduce**

Assuming you have two user accounts with `Neos.Neos:Editor` role (let's call them "editor-a" and "editor-b"):

1. Log in as "editor-a"
2. Create a document "Test"
3. Add a text node to "Test"
4. Publish those changes
5. Edit the text node, without publishing that change
6. Log in as "editor-b" in a private window (keep "editor-a"'s window open!)
7. Sync "editor-b"'s workspace, so you can see document "Test"
8. Remove document "Test"
9. Publish that change
10. Switch to "editor-a"'s window (do not reload the page!)
11. Try to publish the change on the text node that is still pending
12. Observe:
  - **Pre-PR:** Publishing results in an exception
  - **Post-PR:** Publishing triggers conflict resolution

### The solution

With this PR, the UI automatically triggers the sync and conflict resolution workflows if a publishing operation fails due to an outdated workspace.

### Remaining TODOs

- [x] Move `Conflicts` concept to `Application\Shared` namespace
  - [x] Move `ConflictsOccurred` to `Application\Shared` namespace
  - [x] Move `Conflict` to `Application\Shared` namespace
  - [x] Move `Conflicts` to `Application\Shared` namespace
  - [x] Move `ConflictsBuilder` to `Application\Shared` namespace
  - [x] Move `ReasonForConflict` to `Application\Shared` namespace
  - [x] Move `TypeOfChange` to `Application\Shared` namespace
  - [x] Move `IconLabel` to `Application\Shared` namespace
- [x] Add command handlers for all Publish-related commands & make sure they throw `ConflictsOccurred`
  - [x] Move `PublishChangesInDocument` -> `PublishChangesInDocument\PublishChangesInDocumentCommand`
  - [x] Create `PublishChangesInDocumentCommandHandler`
  - [x] Ensure that `PublishChangesInDocumentCommandHandler` throws `ConflictsOccurred`
  - [x] Move `PublishChangesInSite` -> `PublishChangesInSite\PublishChangesInSiteCommand`
  - [x] Create `PublishChangesInSiteCommandHandler`
  - [x] Ensure that `PublishChangesInSiteCommandHandler` throws `ConflictsOccurred`
- [x] Turn `ConflictsOccurred` into a Result DTO rather than an exception
- [x] Eliminate redundancy in `catch (WorkspaceRebaseFailed $e)` clauses
- [x] Trigger conflict resolution from within `Publish` saga 
  - [x] Add `conflicts` case to `PublishingResponse`
  - [x] Add `PublishingState.CONFLICTS`
- [x] Fix: "Node could not be published, because of missing parentNode" after conflict resolution (happens when you do "Publish Document" on a document that gets removed by conflict resolution)
- [x] Fix: TypeError node.children is undefined (regression after https://github.com/neos/neos-ui/pull/3756 - Race Condition)
- [x] Fix: You still can get stuck in an undefined state in which the saga ceases to continue (happens when the in-between sync is cancelled)
- [x] Fix: #3785

### CODE REVIEW

**Comment by mhsdesign** [comment]
Id like to help reviewing this so we can soon make the changes Neos requires regarding the rename an removal of `Node` fields on 9.0-dev (whithout major conflicts against your pr)

**Comment by grebaldi** [comment]
@mhsdesign Don't worry about conflicts - I'll take care of keeping this PR up-to-date. So, unless anything upstream of the UI needs this change to move on, I'd say that everything else has more priority.

**Comment by grebaldi** [comment]
@pKallert 
FYI: While rebasing, I encountered conflicts with the fix you've provided in https://github.com/neos/neos-ui/pull/3634. To resolve this, I have moved the translation handling from `BackendServiceController::publishChangesInDocumentAction` to `PublishChangesInDocumentCommandHandler::handle`:
https://github.com/neos/neos-ui/blob/faee8cad95b72ad96e2541a0b40e09d28645393b/Classes/Application/PublishChangesInDocument/PublishChangesInDocumentCommandHandler.php#L62-L74 

Does that match your intend?

@mhsdesign 
One of the conflicting commits was one of yours, but was introduced in #3634 as well: https://github.com/neos/neos-ui/commit/29a73a166afe1f3e2a545a577494b62875262785

So, same question goes to you as well :)

**Comment by mhsdesign** [comment]
Yes looks perfect thanks for taking care :)

**Comment by mhsdesign** [comment]
Hi i kindof forget about this ^^ Would this be something for the next beta to include?

And while navigating the code i found that this file is no longer in use `packages/neos-ui-redux-store/src/UI/SyncWorkspaceModal/index.ts` so we might as well delete that? Seems its a leftover from paulas previous pr?

File: Classes/Infrastructure/ContentRepository/ConflictsFactory.php [hunk 1]
```diff
@@ -127,6 +114,9 @@ private function createConflictFromCommandThatFailedDuringRebase(
             : null;
 
         return new Conflict(
+            key: $affectedNode
+                ? $affectedNode->nodeAggregateId->value
+                : 'command-' . $commandThatFailedDuringRebase->sequenceNumber,
```
Comment (mhsdesign) [inline Classes/Infrastructure/ContentRepository/ConflictsFactory.php]: this is some kind of deduplication thing i think but i dont fully understand it yet :D

File: Classes/Infrastructure/ContentRepository/ConflictsFactory.php [hunk 2]
```diff
@@ -32,72 +32,59 @@
...
-        );
+        $this->workspace = $contentRepository->getWorkspaceFinder()
+            ->findOneByName($workspaceName);
+    }
 
-        $this->items[] = $conflict;
+    public function fromWorkspaceRebaseFailed(
+        WorkspaceRebaseFailed $workspaceRebaseFailed
+    ): Conflicts {
+        $conflictsBuilder = Conflicts::builder();
```
Comment (mhsdesign) [inline Classes/Infrastructure/ContentRepository/ConflictsFactory.php]: i know you like builders, but id be fine by using and array to track it ^^

File: Classes/Infrastructure/ContentRepository/ConflictsFactory.php [hunk 3]
```diff
@@ -127,6 +114,9 @@ private function createConflictFromCommandThatFailedDuringRebase(
             : null;
 
         return new Conflict(
+            key: $affectedNode
+                ? $affectedNode->nodeAggregateId->value
```
Comment (mhsdesign) [inline Classes/Infrastructure/ContentRepository/ConflictsFactory.php]:
```suggestion
                ? $affectedNode->aggregateId->value
```

**Review by mhsdesign — COMMENTED** [review]
There where some changes in the e2e setup and the test dont pass all on 9.0 anymore but that might be fixed here as well. But i still happen to get publish errors when running the tests locally: And the tests fail at a later point because we just click the `Discard` button with the id `#neos-DiscardDialog-Acknowledg` even if its an error:

image

**Comment by mhsdesign** [comment]
That rebasing doesnt work is actually a core bug and will be resolved via: https://github.com/neos/neos-development-collection/pull/5301 ... that fixes most of the tests locally but one is still failing ...

**Review by mhsdesign — APPROVED** [review]
Thank you so much ❤️ tested in combination with the new publishing v3 everything works again locally: https://github.com/neos/neos-development-collection/pull/5301

(I solved conflicts with 9.0 and also pushed a few minor changes)

**Pull request merged**